src/ir_passes/await.ml

```diff
@@ -37,6 +37,19 @@ let letcont k scope =
     blockE [funcD k' v e] (* at this point, I'm really worried about variable capture *)
             (scope k')
 
+(* pre-compose a continuation with a call to a `finally`-thunk *)
+let precont k thunk =
+  let finally = blockE [expD (thunk -*- unitE ())] in
+  match k with
+  | ContVar k' ->
+     let typ = match typ_of_var k' with
+       | T.(Func (Local, Returns, [], [], _)) -> T.unit
+       | T.(Func (Local, Returns, [], [typ], _)) -> typ
+       | _ -> assert false in
+    MetaCont (typ, fun v -> finally (varE k' -*- varE v))
```

Comment:
am I careful enough avoiding code duplication here?

Reply:
by using `meta` instead of `MetaCont` directly, we'll assert non-duplication

Reply:
Indeed, I get assertion failures when doing it. Investigating.

Reply:
Did you resolve this?

Reply:
No, it is on the TODO list (in the description). I have no idea how to fix it (not a semantic problem) and I prefer to write a follow-on PR to handle it.

Reply:
Hmm, that's a wee bit worrying.